libs/langchain/langchain/vectorstores/pinecone.py

```diff
@@ -108,20 +118,32 @@ def add_texts(
             List of ids from adding the texts into the vectorstore.
 
         """
+        # TODO: Parametrize CHUNK_SIZE
+        CHUNK_SIZE = 1000
         if namespace is None:
             namespace = self._namespace
-        # Embed and create the documents
-        docs = []
         ids = ids or [str(uuid.uuid4()) for _ in texts]
-        embeddings = self._embed_documents(texts)
-        for i, (text, embedding) in enumerate(zip(texts, embeddings)):
-            metadata = metadatas[i] if metadatas else {}
+        metadatas = metadatas or [{} for _ in texts]
+        for index, text in enumerate(texts):
+            metadata = metadatas[index]
             metadata[self._text_key] = text
-            docs.append((ids[i], embedding, metadata))
-        # upsert to Pinecone
-        self._index.upsert(
-            vectors=docs, namespace=namespace, batch_size=batch_size, **kwargs
-        )
+        # TODO: add tenacity retry logic
+        for chunk in chunks(zip(ids, texts, metadatas), batch_size=CHUNK_SIZE):
+            chunk_ids, chunk_texts, chunk_metadatas = zip(*chunk)
+            embeddings = self._embed_documents(chunk_texts)
+            async_res = [
+                self._index.upsert(
+                    vectors=list(batch),
+                    namespace=namespace,
+                    **(kwargs or {}),
+                    async_req=True,
+                )
+                for batch in chunks(
+                    zip(chunk_ids, embeddings, chunk_metadatas), batch_size=batch_size
+                )
+            ]
+            [res.get() for res in async_res]
```

Comment:
NIT: add a comment above explaining what we're doing here and why we are doing it in this way.